Split Ansible docs from core docs

samccann opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

-  creates separate index and conf.py files (for ansible, 2.10,  and core) and symlinks in the appropriate files during make.
    - Ansible files are needed for Ansible 3 and future, but must be build from stable-2.10.
   - 2.10 files are needed specifically for Ansible 2.10, which must be buildable in stable-2.10 for some time to come.
   - core files - needed to build the subset of docs to a new url for core - docs.ansible.com/ansible-core/<version>

 - creates a make coredocs command to make only...core docs (no scenario or network guides and builtin collection only for example.

 - splits up the porting guides and roadmaps so only the appropriate files show up on their respective docsites.

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
Partially addresses #72032
##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Docs Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
docs.ansible.com
##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

samccann committed:
create separate conf.py files and test exclusion

samccann committed:
add new index for ansible the package

samccann committed:
exclude scenario guides from core docs

abadger committed:
Pass the path to the sphinx config file.

This way we don't have to maintain two copies of the Makefile.

samccann committed:
use only directive instead of separate index files

samccann committed:
clean up index-works for both

samccann committed:
add make coredocs to makefiles

samccann committed:
remove tag

samccann committed:
switch back to old index style, with 2 index files

samccann committed:
remove only directive

samccann committed:
remove dummy edits

samccann committed:
fix versions

samccann committed:
fix ci error

samccann committed:
split porting guides

samccann committed:
separate roadmaps

samccann committed:
simplify porting guide approach

samccann committed:
symlink indices to create index.html

abadger committed:
Add flag to specify ansible version for the docs build.

When building split docs, we'll need to be able to aspecify which
version of ansible we're building for.

abadger committed:
Add method to build 2.10 docs as well.

We want to build three types of docs:
* ansible-2.10 and earlier: all the docs.  Handle this via `make webdocs
  ANSIBLE_VERSION=2.10`
* ansible-3 and later: a subset of the docs for the ansible package.
  Handle this via `make webdocs ANSIBLE_VERSION=3` (change the
  ANSIBLE_VERSION to match the version being built for.
* ansible-core: a subset of the docs for the ansible-core package.
  Handle thhs via `make coredocs`.

The implementations uses a separate index.rst for each of the three
cases so that the menu displays the right entries and a separate conf.py
for each case which excludes the docs which we don't want built for that
case.  The proper files are symlinked into place.

Two related changes:
* `make webdocs` now always builds all the collection docs.  Use `make
  coredocs` to limit it to core plugins only
* If no ANSIBLE_VERSION is specified, build plugins for the latest release of ansible

abadger committed:
Have the decision to build all docs use the version specified by the user

I was using the branch's version but it makes more sense to use the
version the user directly specified.

samccann committed:
fix build warnings

abadger committed:
Revert "Have the decision to build all docs use the version specified by the user"

This reverts commit 64663e008a78004fd41f6daf5b9f64b82423b144.

In jenkins, we'll have branch information but not ANSIBLE_VERSION so
branch is the better choice after all.

samccann committed:
clean build warnings

samccann committed:
fix docs-build sanity test error

webknjaz commented on docs/docsite/Makefile:
```diff
@@ -50,23 +59,47 @@ all: docs
+	# script so use it for version comparison
+	if python -c "import sys, packaging.version as p; sys.exit(not p.Version('$(MAJOR_VERSION)') >  p.Version('2.10'))" ; then \
+		echo "Creating symlinks in generate_rst"; \
+		ln -sf ../rst/ansible_index.rst rst/index.rst; \
+		ln -sf ../sphinx_conf/ansible_conf.py rst/conf.py; \
+	else \
+		echo 'Creating symlinks for older ansible in generate_rst'; \
+		ln -sf ../rst/2.10_index.rst rst/index.rst; \
+		ln -sf ../sphinx_conf/2.10_conf.py rst/conf.py; \
+	fi
+
+# Symlink the relevant index into place for building core docs
+core_structure: core_generate_rst
+	@echo "Creating symlinks in core_generate_rst"
+	-ln -sf ../rst/core_index.rst rst/index.rst
+	-ln -sf ../sphinx_conf/core_conf.py rst/conf.py
```
I don't think symlinking is necessary. Just pass different configs and settings via CLI args.
Look:
```console
$ sphinx-build --help
usage: sphinx-build [OPTIONS] SOURCEDIR OUTPUTDIR [FILENAMES...]

[...]


build configuration options:
  -c PATH           path where configuration file (conf.py) is located
                    (default: same as SOURCEDIR)
[...]
  -D setting=value  override a setting in configuration file
[...]
```

So you could invoke something like
```console
$ python -m sphinx -c sphinx_conf_dir_2.10/ -D master_doc=ansible_index ...
```
and there would be no need to mutate files on disk.

webknjaz replied:
Also, we could explore if using tags (https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#tags) for conditional ToCs...

samccann replied:
fwiw - both were tried and rejected by others in favor of the symlink approach. We'll see what other comments there are.

mattclay replied:
These symlinks appear to be responsible for the package-data sanity test failure. Instead of linking them, have you tried copying them instead?

samccann replied:
Tried the copy instead of symlink localy and that does seem to work and removes the build errors, thanks!

abadger replied:
Hmmm... you wouldn't want the files (the copies inside of the docs/docsite/rst directory) to end up in the sdist, though, as they're generated files.  Does the build error go away because they do get added to the sdist if they're copied?  If so, it's not the correct fix.  Instead we probably want to add an exclude for those files to MANIFEST.in

mattclay replied:
In that case we'll need to ignore them in the sanity test instead of copying them, by adding them here:

https://github.com/ansible/ansible/blob/00bd0b893d5d21de040b53032c466707bacb3b93/test/sanity/code-smell/package-data.py#L38

abadger replied:
I've got this working.  For posterity, there are several locations that generated files may need to be listed:

* `clean target` of a Makefile so that they're removed when make clean is run
* `MANIFEST.in` they should normally be excluded in here so they do not end up in the tarball
* `.gitignore` they should be added here so that they aren't checked in to the git repository
* `test/sanity/code-smell/package-data.py` they may need to be added to an ignore list here if we don't want them in the tarball and they aren't ignored by an existing ignore pattern.

webknjaz replied:
@samccann what was the reason for not pointing to separate configs?

samccann replied:
@webknjaz - I'm not sure at this point. I remember I hit a roadblock and had to have @abadger help me out of it. As part of that help, we went to symlinks instead of  the separate -c sphinx option and sphinx tags.  I can try it again post-merge and see if I can get it to work. At this point, we have something that works and we need it merged today/backported asap to get the release out the door.

webknjaz replied:
@abadger I'd be interested in hearing the details. Over the past year or so I've learned a lot about Sphinx and making Sphinx extensions so maybe I'd be able to help out.

acozine replied:
I think if we're going to leave the version-switcher in place, we need to keep `index.html` as the name of the default main page.

mattclay commented on Makefile:
```diff
@@ -5,6 +5,7 @@
 # useful targets:
 #   make clean ---------------- clean up
 #   make webdocs -------------- produce ansible doc at docs/docsite/_build/html
+#   make coredocs -------------- produce core doc at docs/docsite/_build/html
```
```suggestion
#   make coredocs ------------- produce core doc at docs/docsite/_build/html
```

abadger committed:
Various ignores for the generated files.

conf.py and the toplevel index.rst are now generated by the docs build
so we don't want them included in the git repo or included in the
tarball.

abadger committed:
Update Makefile

Co-authored-by: Matt Clay <matt@mystile.com>

samccann committed:
update index

samccann commented:
The details of this implementation for future reference:
 excludes scenario guides from core docs, splits porting guides and roadmaps, symlinks indices to create index.html pages, and adds .gitignore entries for conf.py and the toplevel index.rst files generated by the docs build
This solution builds three types of docs:
* ansible-2.10 and earlier: all the docs.  Handle this via `make webdocs
  ANSIBLE_VERSION=2.10`
* ansible-3 and later: a subset of the docs for the ansible package.
  Handle this via `make webdocs ANSIBLE_VERSION=3` (change the
  ANSIBLE_VERSION to match the version being built for.
* ansible-core: a subset of the docs for the ansible-core package.
  Handle this via `make coredocs`.
* `make webdocs` now always builds all the collection docs
*  Use `make coredocs` to limit it to core plugins only
*  The user specifies the desired version. If no ANSIBLE_VERSION is specified, build plugins for the latest release of ansible

webknjaz commented on docs/docsite/rst/ansible_index.rst:
```diff
@@ -0,0 +1,104 @@
+   :maxdepth: 2
+   :caption: Roadmaps
+
+   roadmap/ansible_roadmap_index.rst
```
Looks like this should've also linked `ansible_base_roadmap_index`...
